### PR TITLE
fix(ivorysql_ora): return IEEE results from binary float arithmetic

### LINKED ISSUE #1737: BINARY_FLOAT/BINARY_DOUBLE arithmetic raises on overflow, underflow and division by zero
## Bug Report

Dividing a `BINARY_FLOAT` or `BINARY_DOUBLE` by zero raises an error. Oracle
returns Infinity. Overflow and underflow are the same story. I hit this while
diffing IvorySQL against Oracle for a migration project. #375 fixed the
special-value *literals*; arithmetic that *produces* those values still errors
out.

### IvorySQL Version

IvorySQL 3.4 (PostgreSQL 16.4), built from source.
Also reproduced on master, IvorySQL 5beta1 (PostgreSQL 19devel).

### OS Version (uname -a)

Linux 5.14.0-611.41.1.el9_7.x86_64 #1 SMP PREEMPT_DYNAMIC x86_64 GNU/Linux (Rocky 9)

### Configuration options ( config.status --config )

'--prefix=/home/muzzammil/ivory16-install'

### Current Behavior

| Expression | IvorySQL 3.4 and master |
|---|---|
| `1d / 0d` | `ERROR 22012 division by zero` |
| `-1d / 0d` | `ERROR 22012 division by zero` |
| `0d / 0d` | `ERROR 22012 division by zero` |
| `1f / 0f` | `ERROR 22012 division by zero` |
| `1e308d * 10` | `ERROR 22003 value out of range: overflow` |
| `-1e308d * 10` | `ERROR 22003 value out of range: overflow` |
| `1e308d + 1e308d` | `ERROR 22003 value out of range: overflow` |
| `3.4e38f * 10` | `ERROR 22003 value out of range: overflow` |
| `1e-320d / 1e10` | `ERROR 22003 value out of range: underflow` |
| `1e-45f / 1e10` | `ERROR 22003 value out of range: underflow` |

Normal arithmetic is fine: `2d * 3d` = 6, `7d / 2d` = 3.5, `1f + 1d` = 2.

### Expected behavior/code

Same expressions on Oracle Database 21c Express Edition (21.3.0.0.0):

| Expression | Oracle 21c |
|---|---|
| `1d / 0d` | `inf` |
| `-1d / 0d` | `-inf` |
| `0d / 0d` | `nan` |
| `1f / 0f` | `inf` |
| `1e308d * 10` | `inf` |
| `-1e308d * 10` | `-inf` |
| `1e308d + 1e308d` | `inf` |
| `3.4e38f * 10` | `inf` |
| `1e-320d / 1e10` | `0` |
| `1e-45f / 1e10` | `0` |

Oracle documents both types as IEEE 754 binary32 and binary64. Under IEEE,
division by zero gives +/-Infinity (NaN for 0/0), overflow gives +/-Infinity,
and underflow gives zero. None of it raises.

To be clear, Oracle's `NUMBER` does raise on overflow (`ORA-01426`), so this
isn't an "Oracle never raises" claim. It's specific to the two binary
floating-point types.

Division by zero is the one I'd actually worry about. A divisor that turns out
to be zero is ordinary application data. You need fairly extreme values to reach
the overflow and underflow cases.

### Step to reproduce

```sql
SET ivorysql.compatible_mode = oracle;

SELECT 1d / 0d;          -- expected inf,  got ERROR 22012 division by zero
SELECT 0d / 0d;          -- expected nan,  got ERROR 22012 division by zero
SELECT 1e308d * 10;      -- expected inf,  got ERROR 22003 overflow
SELECT 3.4e38f * 10;     -- expected inf,  got ERROR 22003 overflow
SELECT 1e-320d / 1e10;   -- expected 0,    got ERROR 22003 underflow
```

The `SET` has to go in its own statement. If it shares a simple-query batch with
the SELECT, the whole batch is parsed before anything executes, so the SELECT is
still parsed in `pg` mode. Took me a while to work that one out.

### Additional context that can be helpful for identifying the problem

The operators for these types map straight onto PostgreSQL's internal float
functions, in `contrib/ivorysql_ora/src/datatype/datatype--1.0.sql`:

```sql
CREATE FUNCTION sys.binary_float_mul(sys.binary_float, sys.binary_float)
RETURNS sys.binary_float
AS 'float4mul'
LANGUAGE internal
```

Which lands in the inline helpers in `src/include/utils/float.h`:

```c
result = val1 * val2;
if (unlikely(isinf(result)) && !isinf(val1) && !isinf(val2))
    float_overflow_error();
if (unlikely(result == 0.0) && val1 != 0.0 && val2 != 0.0)
    float_underflow_error();
```

PostgreSQL raising there is deliberate, and it's a fair call for its own
`float4`/`float8`. An Infinity quietly landing in an aggregate is easy to miss.
I'm not arguing with that. What I'm unsure about is whether it should carry over
to types whose whole job is to behave like Oracle's.

One thing I couldn't work out:
`contrib/ivorysql_ora/expected/ora_binary_double.out` currently has

```
SELECT '' AS bad, f.f1 * '1e200' from BINARY_DOUBLE_TBL f;
ERROR:  value out of range: overflow
```

so the suite passes as it stands. I can't tell whether that's a deliberate
decision to keep PostgreSQL semantics here, or whether the file came across from
PostgreSQL's `float8` tests.

Three questions before I take this any further:

1. Is the current behaviour intentional for `BINARY_FLOAT` and `BINARY_DOUBLE`?
2. If not, should a fix key off `ivorysql.compatible_mode`, IEEE semantics under
   `oracle` and current behaviour under `pg`, so nothing relying on the error
   breaks?
3. More generally, is there a project position on what wins when Oracle and
   PostgreSQL semantics conflict on an Oracle-compatible type? I'd rather learn
   the rule than ask case by case.

I have this working locally already: new arithmetic functions in the extension,
operators rebound, no PostgreSQL core files touched, plus regression cases built
from the Oracle output above. Happy to turn it into a PR, but I'd rather know
which behaviour you want first.

Unrelated: the contribution guidelines mention signing the CLA, but I couldn't
find where the signed copy is meant to go. What's the right route?

### PR DESCRIPTION
**Pull request by Muzzammil242**
## Summary

Arithmetic on `BINARY_FLOAT` and `BINARY_DOUBLE` raises on overflow, underflow and division by zero. Oracle returns `Inf`, `0` and `NaN` for those cases — both types are documented as IEEE 754 binary32 and binary64.

This adds arithmetic functions to the extension that return the IEEE result, and rebinds the sixteen operators to them.

Fixes #1737

## Root cause

The operators were bound to PostgreSQL's internal float functions — `float4mul`, `float8pl` and the rest. Those call the inline helpers in `src/include/utils/float.h`, which compute the IEEE result and then throw it away:

```c
result = val1 * val2;
if (unlikely(isinf(result)) && !isinf(val1) && !isinf(val2))
    float_overflow_error();
```

That is the right choice for `float4` and `float8`, which never promised IEEE exception semantics. It is the wrong one for two types that exist to model Oracle's.

Patching `float.h` behind a GUC was the obvious alternative, and I did not take it. `float.h` is PostgreSQL's: changing it moves `float4`/`float8` behaviour for every PostgreSQL user of this fork, in a direction PostgreSQL chose deliberately. It would also put a GUC check in the hottest arithmetic path in the backend, and leave a conflict to resolve in a core header on every PostgreSQL merge. `ivorysql_ora` already owns both types, so the operators are rebound there instead.

Comparison operators are left alone. Equality and ordering have no exception semantics to get wrong.

## Reproduction

Oracle Database 21c Express Edition, against IvorySQL master with `ivorysql.compatible_mode = oracle`:

```sql
SELECT 3.4e38f * 10f;   -- Oracle: Inf   IvorySQL: ERROR 22003 value out of range
SELECT 1e-45f / 1e10;   -- Oracle: 0     IvorySQL: ERROR 22003 value out of range
SELECT 1f / 0f;         -- Oracle: Inf   IvorySQL: ERROR 22012 division by zero
SELECT 0f / 0f;         -- Oracle: NaN   IvorySQL: ERROR 22012 division by zero
```

The underflow case is easy to miss. The guard in `float.h` is skipped when either operand is zero, so `1e-320d * 0.0` returns `0` correctly while `1e-320d / 1e10` raises.

## Testing

`contrib/ivorysql_ora` regression on master, same tree, patch applied and reverted between runs:

```
stock master      1 of 26 failed (ora_xml_functions)
with this patch   1 of 26 failed (ora_xml_functions)

ok 5   - ora_binary_float
ok 6   - ora_binary_double
```

Expected output is updated for the cases this changes. `ora_xml_functions` fails on stock master and is unrelated.

Rocky Linux 9, gcc 11.5, `--enable-depend --with-uuid=e2fs`.

---

Assisted-by: Claude:Opus-5
Percentage of AI-generated code: 95%

## Summary by CodeRabbit

- **New Features**
  - Added Oracle-compatible IEEE 754 arithmetic for `BINARY_FLOAT` and `BINARY_DOUBLE`.
  - Overflow and underflow now produce Infinity or zero as appropriate.
  - Division by zero returns `Infinity`, `-Infinity`, or `NaN` instead of an error.
  - Mixed-precision arithmetic promotes values consistently to `BINARY_DOUBLE`.
  - Ordinary arithmetic behavior remains unchanged.

- **Tests**
  - Added coverage for overflow, underflow, division by zero, ordinary arithmetic, signed results, and precision promotion.

### CODE REVIEW

**Comment by Muzzammil242** [comment]
Extended the mixed-precision coverage, which was thin: the patch rebinds eight mixed operators and the tests exercised one of them.

Added both operand orders for `+`, `-`, `*` and `/` with `pg_typeof` assertions, since a value test alone does not show the result type — `1f + 1d` is `2` whether the result is `binary_float` or `binary_double`.

Also added two range cases and one precision case that fail on the value alone if the type is wrong: `3.4e38f * 10d` overflows as binary32, and `1f + 1e-10d` rounds to exactly `1` as binary32. Values checked against Oracle 21c.

Regression on master is unchanged: 1 of 26 failed before and after, `ora_xml_functions`, which fails on stock master and is unrelated.

**Comment by hs-liuxh** [comment]
Thanks for your contribution.

File: contrib/ivorysql_ora/sql/ora_binary_float.sql [hunk 1]
```diff
@@ -192,3 +191,37 @@ SELECT to_binary_float('-inf');
...
+-- Database 21c Express Edition.
+--
+
+-- overflow and underflow
+SELECT 3.4e38f * 10;
+SELECT 1e-45f / 1e10;
+
+-- division by zero
+SELECT 1f / 0f;
+SELECT 0f / 0f;
```
[the remark this answers is not in the document]
Comment (Muzzammil242) [reply contrib/ivorysql_ora/sql/ora_binary_float.sql:213]: Thanks for catching this, the signed cases were genuinely missing. I've added them, but not in exactly the form suggested. Two reasons.

First, `(-3.4e38f) * 10` doesn't overflow. Unary minus on these types comes back as float8:

```sql
SELECT pg_typeof(-1f), pg_typeof(-1d);
-- pg_catalog.float8 | pg_catalog.float8

SELECT (-3.4e38f) * 10;
-- -3.399999952144364e+39
```

So the expression drops out of binary_float arithmetic altogether and returns an ordinary float8. It gives that same answer with and without this patch, so it wouldn't catch the regression it's meant to catch. I used a subtraction from zero instead, which keeps the type: `pg_typeof(0f - 3.4e38f)` is `binary_float`, and on Oracle `0f - 3.4e38f` and `-3.4e38f` are the same bits (9ec97fff), so nothing is lost by writing it that way.

Second, `1f / (-0f)` returns `Inf` on Oracle, not `-Infinity`. Oracle doesn't keep a negative zero in BINARY_FLOAT or BINARY_DOUBLE. I checked on 21c XE a few different ways, dividing by the zero so the sign shows up in the result rather than trusting how the client prints it:

```
SELECT 1f / -0f FROM dual;                    -- Inf
SELECT 1f / to_binary_float('-0') FROM dual;  -- Inf
SELECT 1f / (0f * (0f - 1f)) FROM dual;       -- Inf
SELECT 1d / -0d FROM dual;                    -- Inf
```

IvorySQL returns `-Inf` for that expression, so there is a real difference from Oracle here, but it is about how signed zero is represented rather than about the arithmetic this patch changes. If I added the case with the `-Inf` we currently produce, the expected file would be recording behaviour Oracle doesn't have. I'd rather raise that one separately.

So what went in is three cases, each checked against Oracle 21c and each failing on master without the patch:

```
SELECT (0f - 3.4e38f) * 10;       -- Oracle -Inf;  master: ERROR value out of range: overflow
SELECT (0f - 3.4e38f) - 3.4e38f;  -- Oracle -Inf;  master: ERROR value out of range: overflow
SELECT (0f - 1f) / 0f;            -- Oracle -Inf;  master: ERROR division by zero
```

One more thing: `ora_binary_double.sql` had `-1d / 0d` for the same case. I've changed it to `(0d - 1d) / 0d` so both files use one form. The result is unchanged.

File: contrib/ivorysql_ora/sql/ora_binary_float.sql [hunk 2]
```diff
@@ -192,3 +191,46 @@ SELECT to_binary_float('-inf');
...
+
+-- Signed results. The sign of an operand has to survive into the exception
+-- value: a rule that returned +Inf for these would still pass every case
+-- above. Written as a subtraction from zero rather than as unary minus,
+-- which resolves to float8 on these types and would take the expression out
+-- of binary_float arithmetic altogether.
+SELECT (0f - 3.4e38f) * 10;
+SELECT (0f - 3.4e38f) - 3.4e38f;
+SELECT (0f - 1f) / 0f;
```
[the remark this answers is not in the document]
Comment (Muzzammil242) [reply contrib/ivorysql_ora/sql/ora_binary_float.sql:222]: This one crossed with my comment on the thread above, posted a few seconds apart.

It's the same expression I measured there. Oracle returns `Inf` for it, not `-Inf`:

```
SELECT 1f / (0f * (0f - 1f)) FROM dual;  -- Inf
SELECT 1f / -0f FROM dual;               -- Inf
SELECT 1d / -0d FROM dual;               -- Inf
```

Oracle doesn't keep a negative zero in BINARY_FLOAT or BINARY_DOUBLE, so a negative-zero divisor divides as `+0`. We do return `-Inf` there, so there is a real difference from Oracle, but it comes from how signed zero is represented rather than from the arithmetic this patch changes. Adding the case with `-Inf` would put behaviour Oracle doesn't have into the expected file, so I'm leaving it for a separate issue.

**Comment by hs-liuxh** [comment]
This PR is crucial for ensuring IvorySQL's behavior remains consistent with Oracle. Thanks for your contribution! 

Before we can merge it, please add test cases to cover the following two missing scenarios: negative sign underflow and underflow via multiplication. For examples:
```sql
SELECT (0d - 1e-320d) / 1e10; 
SELECT (0f - 1e-45f) / 1e10;

SELECT 1e-200d * 1e-200d;
SELECT 1e-20f * 1e-30f;
```

**Comment by Muzzammil242** [comment]
Thanks for the review. All four cases are added, in both test files, with one finding along the way that's worth spelling out.

Your two negative-underflow cases are exactly where IEEE and Oracle part ways. Under pure IEEE 754, the underflow of a negative value produces negative zero, so `(0f - 1e-45f) / 1e10` is `-0`. Oracle doesn't keep a negative zero in these types. I measured on 21c, dividing one by each result so the zero's sign shows up as the sign of the infinity rather than depending on how a client prints zero:

```
SELECT (0d - 1e-320d) / 1e10 FROM dual;        -- 0
SELECT (0f - 1e-45f) / 1e10 FROM dual;         -- 0
SELECT 1e-200d * 1e-200d FROM dual;            -- 0
SELECT 1e-20f * 1e-30f FROM dual;              -- 0

SELECT 1d / ((0d - 1e-320d) / 1e10) FROM dual; -- Inf, so the zero is +0
SELECT 1f / ((0f - 1e-45f) / 1e10) FROM dual;  -- Inf
SELECT 1f / (0f * (0f - 1f)) FROM dual;        -- Inf, same for 0 * negative
```

So the patch now normalises a zero result to +0 in the arithmetic wrappers (a small inline helper in each file, applied to all sixteen functions). With that, all four of your cases return `0` exactly as Oracle does, and the tests would catch an implementation that kept the IEEE `-0`. It's the one deliberate departure from IEEE in the patch, made to follow Oracle - same reasoning as the rest of the change, just pointing the other way.

Regression is unchanged otherwise: no previously expected value moved, and the full suite is at 1 of 26 failed, the pre-existing ora_xml_functions case that needs a --with-libxml build.

The remaining signed-zero surface - `to_binary_float('-0')` coming in through the input function rather than from arithmetic - is a separate path and I'd rather raise it on its own than grow this patch further.
